client/src/main/java/org/evosuite/ga/metaheuristics/mosa/AdaptiveDynaMOSA.java

```diff
@@ -0,0 +1,252 @@
+
+        for (int i = offspringPopulation.size() - 1; i >= 0; i--) {
+            TestChromosome off = offspringPopulation.get(i);
+            if (off.getPerformanceScore() == Double.MAX_VALUE) {
+                offspringPopulation.remove(i);
+            }
+        }
+
+        /* select the heuristic */
+        //logger.error("Last Heuristic = {}", last_heuristic);
+        last_heuristic = checkStagnation();
+        //logger.error("Current Heuristic = {}", last_heuristic);
+
+        // Create the union of parents and offSpring
+        List<TestChromosome> union = new ArrayList<>(this.population.size() + offspringPopulation.size());
+        union.addAll(this.population);
+        union.addAll(offspringPopulation);
+
+        // Ranking the union
+        logger.debug("Union Size = {}", union.size());
+
+        // Ranking the union using the best rank algorithm (modified version of the non dominated sorting algorithm)
+        this.rankingFunction.computeRankingAssignment(union, this.goalsManager.getCurrentGoals());
+
+        // let's form the next population using "preference sorting and non-dominated sorting" on the
+        // updated set of goals
+        int remain = Math.max(Properties.POPULATION, this.rankingFunction.getSubfront(0).size());
+        int index = 0;
+        List<TestChromosome> front;
+        this.population.clear();
+
+        // Obtain the next front
+        front = this.rankingFunction.getSubfront(index);
+
+        while ((remain > 0) && (remain >= front.size()) && !front.isEmpty()) {
+
+            applySecondaryCriterion(front);
+
+            logger.debug("Distance = {}, Score = {} ", front.get(0).getDistance(), front.get(0).getPerformanceScore());
+
+            // Add the individuals of this front
+            this.population.addAll(front);
+
+            // Decrement remain
+            remain = remain - front.size();
+
+            // Obtain the next front
+            index++;
+            if (remain > 0)
+                front = this.rankingFunction.getSubfront(index);
+
+        }
+
+        if (remain > 0 && !front.isEmpty()) {
+
+            applySecondaryCriterion(front);
+
+            logger.debug("Distance = {}, Score ={} ", front.get(0).getDistance(), front.get(0).getPerformanceScore());
+            strategy.sort(front);
+
+            for (int k = 0; k < remain; k++)
+                this.population.add(front.get(k));
+
+        }
+
+        this.currentIteration++;
+//        logger.debug("N. fronts = {}", this.rankingFunction.getNumberOfSubfronts());
+//        logger.debug("1* front size = {}", this.rankingFunction.getSubfront(0).size());
+//        logger.debug("Best Values Map size = {}", goalsManager.getBestValues().size());
+        logger.debug("Covered goals = {}", goalsManager.getCoveredGoals().size());
+        logger.debug("Current goals = {}", goalsManager.getCurrentGoals().size());
+        logger.debug("Uncovered goals = {}", goalsManager.getUncoveredGoals().size());
```

Comment:
Can be condensed to 

`
logger.debug("Covered goals = {}, Current goals = {} and Uncovered goals = {}", goalsManager.getCoveredGoals().size(),   goalsManager.getCurrentGoals().size(), goalsManager.getUncoveredGoals().size())
`